src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -1165,6 +1254,143 @@ private function step_in_body() {
 				$this->insert_html_element( $this->state->current_token );
 				return true;
 
+			case '+PLAINTEXT':
+				if ( $this->state->stack_of_open_elements->has_p_in_button_scope() ) {
+					$this->close_a_p_element();
+				}
+
+				/*
+				 * @todo This may need to be handled in the Tag Processor and turn into
+				 *       a single self-contained tag like TEXTAREA, whose modifiable text
+				 *       is the rest of the input document as plaintext.
+				 */
+				$this->last_error = self::ERROR_UNSUPPORTED;
+				throw new WP_HTML_Unsupported_Exception( 'Cannot process PLAINTEXT elements.' );
+
+			/*
+			 * > A start tag whose tag name is "button"
+			 */
+			case '+BUTTON':
+				if ( $this->state->stack_of_open_elements->has_element_in_scope( 'BUTTON' ) ) {
+					// @todo Indicate a parse error once it's possible. This error does not impact the logic here.
+					$this->generate_implied_end_tags();
+					$this->state->stack_of_open_elements->pop_until( 'BUTTON' );
+				}
+
+				$this->reconstruct_active_formatting_elements();
+				$this->insert_html_element( $this->state->current_token );
+				$this->state->frameset_ok = false;
+
+				return true;
+
+			/*
+			 * > An end tag whose tag name is one of: "address", "article", "aside", "blockquote",
+			 * > "button", "center", "details", "dialog", "dir", "div", "dl", "fieldset",
+			 * > "figcaption", "figure", "footer", "header", "hgroup", "listing", "main",
+			 * > "menu", "nav", "ol", "pre", "search", "section", "summary", "ul"
+			 */
+			case '-ADDRESS':
+			case '-ARTICLE':
+			case '-ASIDE':
+			case '-BLOCKQUOTE':
+			case '-BUTTON':
+			case '-CENTER':
+			case '-DETAILS':
+			case '-DIALOG':
+			case '-DIR':
+			case '-DIV':
+			case '-DL':
+			case '-FIELDSET':
+			case '-FIGCAPTION':
+			case '-FIGURE':
+			case '-FOOTER':
+			case '-HEADER':
+			case '-HGROUP':
+			case '-LISTING':
+			case '-MAIN':
+			case '-MENU':
+			case '-NAV':
+			case '-OL':
+			case '-PRE':
+			case '-SEARCH':
+			case '-SECTION':
+			case '-SUMMARY':
+			case '-UL':
+				if ( ! $this->state->stack_of_open_elements->has_element_in_scope( $token_name ) ) {
+					// @todo Report parse error.
+					// Ignore the token.
+					return $this->step();
+				}
+
+				$this->generate_implied_end_tags();
+				if ( ! $this->state->stack_of_open_elements->current_node_is( $token_name ) ) {
+					// @todo Record parse error: this error doesn't impact parsing.
+				}
+				$this->state->stack_of_open_elements->pop_until( $token_name );
+				return true;
+
+			/*
+			 * > An end tag whose tag name is "form"
+			 */
+			case '-FORM':
```

Comment:
do you remember what behavior was not working correctly?

Reply:
Yes, although **it is working correctly now**. We discussed via another channel. We settled on bailing in a problematic case because the breadcrumbs are broken and it's unclear at this time how to resolve it:

https://github.com/WordPress/wordpress-develop/blob/d61f297fed884f2126a15fcdd4e0135eca2c5101/src/wp-includes/html-api/class-wp-html-processor.php#L1559-L1562

My confusion was that a mis-nested form end tag does not close out everything to the form, but does just take the form off the stack of open elements. When I commented, there was a breadcrumbs and depth issue (now resolved by bailing).

Here's an example of the confusing situation:

```html
<!-- misnested form closer -->
<form>
  <div>
    BODY > FORM > DIV
    </form>
    BODY > FORM > DIV (still)
  </div>
BODY (FORM was closed from inside the DIV)

<!-- properly nested form closer -->
<form>
  <div>
    BODY > FORM > DIV
  </div>
  BODY > FORM
</form>
BODY
```

Reply:
Glad that bookmark update solved these. Perhaps the reason I couldn't find a failing test case in 6.6 is simply that we didn't get into the situations where the bug would present itself.